feat: add scanner and sst skeleton

olizilla opened the pull request:
The `Scanner` lambda does a full table scan of the legacy index table, and sends batches of records to an SQS Queue. Stores it's progress in SSM parameter store so it can resume.

The lambda invokes itself again when it's remaining execution time is less than `MIN_REMAINING_TIME_MS`, as we only get 15mins max lambda execution time.

Invoke it directly with 
```shell
aws lambda invoke --function-name <name> --invocation-type Event
```

Pass `{ TotalSegments: number, Segment: number}` to control the [table scan partition](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html#Scan.ParallelScan)

Abort by setting an ssm string parameter with the name `/migrate-block-index/${Config.STAGE}/stop`. All current invocations for that stage will stop. Delete that key and manually invoke them again to pick up where we left off.

Uses the sst Config defaults to get hold of the current stage form inside the lambda. see: https://docs.sst.dev/config#default-values

Inspired by https://github.com/alexdebrie/serverless-dynamodb-scanner


License: MIT

olizilla committed:
feat: add scanner and sst skeleton

The `Scanner` lambda does a full table scan of the legacy index table, and sends batches of records to an SQS Queue. Ported from https://github.com/alexdebrie/serverless-dynamodb-scanner

Uses the sst Config defaults to get hold of the current stage form inside the lambda.
see: https://docs.sst.dev/config#default-values

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla commented:
Example scanner invocation run

<img width="608" alt="Screenshot 2023-09-04 at 14 13 46" src="https://github.com/web3-storage/migrate-block-index-infra/assets/58871/68eb8d76-cb34-403d-93ad-46e5379cb4b3">

olizilla commented:
Self invoking lambda... (with timeout threshold set to 15mins so it always re-invokes, to demonstrate)

<img width="615" alt="Screenshot 2023-09-04 at 15 02 24" src="https://github.com/web3-storage/migrate-block-index-infra/assets/58871/df6434ca-3774-4bb5-a8de-4be81a53b84c">

olizilla committed:
feat: self invoking lambda and scan partition

- Fixes permissions to allow lambda to reinvoke itself when it's about to run out of execution time
- Let us pass table scan partition segment parameters

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: update package-lock

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

vasco-santos commented on README.md:
```diff
@@ -20,6 +20,27 @@ To work on this codebase **you need**:
+ 
+Stores it's progress in SSM parameter store so it can resume.
+ 
+The lambda invokes itself again when it's remaining execution time is less than `MIN_REMAINING_TIME_MS`, as we only get 15mins max lambda execution time.
```
could we consider running a longer job function ? See https://docs.sst.dev/long-running-jobs

olizilla replied:
i considered it, but it's less clear how we would have the jobs self invoke when they run out of time. We get a max run time of 8hrs which is better than 15 mins, but don't know the apis to call to find out how long it has left to run (could just use clock time tho) or how to invoke a code build job from a code build job.

vasco-santos commented on README.md:
```diff
@@ -20,6 +20,27 @@ To work on this codebase **you need**:
+
+Pass `{ TotalSegments: number, Segment: number}` to control the [table scan partition](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html#Scan.ParallelScan)
+
+With ~858 million records to scan, we will run with 10 segments to complete in about ~3 days.
```
This is cool 👍🏼

vasco-santos commented on packages/functions/src/scanner.ts:
```diff
@@ -0,0 +1,125 @@
+  const Segment = event.Segment ?? 0
+
+  // set via the `bind` config in the sst stack
+  const tableName = Table.srcTable.tableName
+  const queueUrl = Queue.batchQueue.queueUrl
```
Are these types working? From my experience with bind config we need to set them, something along the lines of

```ts
declare module 'sst/node/table' {
  export interface TableResources {
    'piece-store': {
      tableName: string;
    };
    'aggregate-store': {
      tableName: string;
    };
    'inclusion-store': {
      tableName: string;
    };
  }
}
```

olizilla replied:
i ran `npx sst types` and it made them exist for me! I didn't dig any further as it seemed to work. I get autocomplete on the things in the lambda impl

vasco-santos replied:
that's quite nice! I assume if you use typescript project it can get those. Thanks for the tip ❤️

vasco-santos commented on packages/functions/src/scanner.ts:
```diff
@@ -0,0 +1,125 @@
+      break
+    }
+    const records = res.Items.map(i => unmarshall(i))
+    recordCount += await sendToQueue(sqs, queueUrl, records)
```
should we skip if records is empty array?

vasco-santos replied:
From my understanding from the [docs](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_ResponseSyntax), more precisely:

> If LastEvaluatedKey is not empty, it does not necessarily mean that there is more data in the result set. The only way to know when you have reached the end of the result set is when LastEvaluatedKey is empty.

it is possible that we get a page with LastEvaluatedKey that then will have no more data

olizilla replied:
Good catch! Thanks!

vasco-santos reviewed (approved):
LGTM
Some minor feedback provided for you to consider

alanshaw commented on README.md:
````diff
@@ -20,6 +20,27 @@ To work on this codebase **you need**:
+aws lambda invoke --function-name <name> --invocation-type Event
+```
+
+Pass `{ TotalSegments: number, Segment: number}` to control the [table scan partition](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html#Scan.ParallelScan)
````
```suggestion
Pass `{ TotalSegments: number, Segment: number }` to control the [table scan partition](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html#Scan.ParallelScan)
```

alanshaw commented on packages/functions/src/scanner.ts:
```diff
@@ -0,0 +1,125 @@
+ * see: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html#Scan.ParallelScan
+ */
+export async function handler(event: any, context: Context) {
+  console.log(`invoked`, event)
```
```suggestion
  console.log('invoked', event)
```

alanshaw commented on packages/functions/src/scanner.ts:
```diff
@@ -0,0 +1,125 @@
+async function invokeSelf(lambda: LambdaClient, event: any, context: Context) {
+  return lambda.send(new InvokeCommand({
+    FunctionName: context.functionName,
+    InvocationType: 'Event',
```
Would be useful to have a comment somewhere around here that just says `InvocationType: 'Event'` doesn't wait for the lambda to finish executing before returning.

alanshaw commented on packages/functions/src/scanner.ts:
```diff
@@ -0,0 +1,125 @@
+    const res = await ssm.send(new GetParameterCommand({
+      Name: key,
+    }))
+    if (res.Parameter && res.Parameter.Value) {
```
Throw otherwise?

alanshaw reviewed (approved):
Looks great. Perhaps put a `p-retry` around some of those service calls, since there's going to be loads of requests and it's going to take a long time. It'll be frustrating to have to kick it off again in the morning after a failure at midnight.

olizilla committed:
feat: retry calls to external services

add extra retries to external service calls. We want these processes to keep running for a few days.

bump up the MIN_REMAINING_TIME_MS to 5mins. 6mins would be the worse case scenario where we have to wait the full 2 mins for each of 3 service calls, but they all succeed on the last try, which seems unlikely

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: fix STOP command handling

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
fix: treat empty Items and missing Items as the same thing.

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla commented:
@alanshaw this will now retry each external service call 10 times over 1min using p-retry, and has an explicit kill switch by setting the ssm parameter value to `"STOP"`. plz to re-eyes.

olizilla committed:
fix: check for STOP value before updating ssm key

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
docs: document STOP flag

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
docs: README

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
docs: README

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: add global stop key and per partiion stop property

Add support for a global stop key. Create an ssm string parameter of the form `/migrate-block-index/${Config.STAGE}/stop` with any value and it will abort all current invocations for that stage (e.g. prod, staging)

Store a `progress` object per partition with the lastEvaluated object and the recordCount so far, to make it easy to see the total progress.

progress also has a stop property which is set to true when the scan is complete to prevent re-scanning the final partition, or restarting the scan if you manually re-invoke. Idempotent!

If you really want to re-run the entire scan, delete all the progess keys from the parameter store first, for a blank slate.

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: tidy up

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: only emit the properties we need

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: tidy

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>
